dm/master/server.go

```diff
@@ -1461,27 +1462,29 @@ func extractWorkerError(result *pb.ProcessResult) error {
 	return nil
 }
 
-/*
-QueryStatus for worker response
-Source:
-	OperateSource:
-		* StartSource, UpdateSource: sourceID = Source
-		* StopSource: return resp.Result = false && resp.Msg = “worker has not started”.
-		  But if len(resp.SourceStatus.Result) > 0, it means dm-worker has some error in
-		  StopWorker and it should be pushed to users.
-Task:
-	StartTask, UpdateTask: query status and related subTask stage is running
-	OperateTask:
-		* pause: related task status is paused
-		* resume: related task status is running
-		* stop: related task can't be found in worker's result
-Relay:
-	OperateRelay:
-		* pause: related relay status is paused
-		* resume: related relay status is running
-In the above situations, once we find an error in response we should return the error.
-*/
-func (s *Server) waitOperationOk(ctx context.Context, cli *scheduler.Worker, taskName, sourceID string, masterReq interface{}) (*pb.QueryStatusResponse, error) {
+// waitOperationOk calls QueryStatus internally to implement a declarative API. It will determine operation is OK by
+// Source:
+//   OperateSource:
+//     * StartSource, UpdateSource: sourceID = Source
+//     * StopSource: return resp.Result = false && resp.Msg = “worker has not started”.
+// Task:
+//   StartTask, UpdateTask: query status and related subTask stage is running
+//   OperateTask:
+//     * pause: related task status is paused
+//     * resume: related task status is running
+//     * stop: related task can't be found in worker's result
+// Relay:
+//   OperateRelay:
+//     * pause: related relay status is paused
+//     * resume: related relay status is running
+// returns OK, error message of QueryStatusResponse, raw QueryStatusResponse, error that not from QueryStatusResponse.
+func (s *Server) waitOperationOk(
+	ctx context.Context,
+	cli *scheduler.Worker,
+	taskName string,
+	sourceID string,
+	masterReq interface{},
+) (bool, string, *pb.QueryStatusResponse, error) {
```

Comment:
Doesn't `*pb.QueryStatusResponse` contain `result` and `msg`? Can you put `OK` and `error message of QueryStatusResponse` in the `pb.QueryStatusResponse`?

https://github.com/pingcap/dm/blob/master/dm/pb/dmworker.pb.go#L407-L414
```
// QueryStatusResponse represents status response for query on a dm-worker
// status: dm-worker's current sub tasks' status
type QueryStatusResponse struct {
	Result        bool             `protobuf:"varint,1,opt,name=result,proto3" json:"result,omitempty"`
	Msg           string           `protobuf:"bytes,2,opt,name=msg,proto3" json:"msg,omitempty"`
	SourceStatus  *SourceStatus    `protobuf:"bytes,3,opt,name=sourceStatus,proto3" json:"sourceStatus,omitempty"`
	SubTaskStatus []*SubTaskStatus `protobuf:"bytes,4,rep,name=subTaskStatus,proto3" json:"subTaskStatus,omitempty"`
}
```

Reply:
the request is not QueryStatus, we use QueryStatus to implement declarative API for other commands, like pause-task (OperateTask)

Reply:
How about directly returning `*pb.CommonWorkerResponse`?

Reply:
that's what caller did in my memory. will check later

Reply:
i think it is ok to return `QueryStatusResponse ` here because what `waitOperationOk` does is send `query-status` request to worker